include/common/network_interface.h

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <string>
+#include <memory>
+#include <functional>
+#include <spdlog/spdlog.h>
+#include <grpc/grpc.h>
+#include <grpcpp/channel.h>
+#include <grpcpp/client_context.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/security/credentials.h>
+#include <core/provider.pb.h>
+#include <core/core.grpc.pb.h>
+
+#include "types.h"
+
+namespace SailGame { namespace Common {
+
+using grpc::Channel;
+using grpc::ClientContext;
+using grpc::ClientReader;
+using grpc::ClientReaderWriter;
+using grpc::ClientReaderWriterInterface;
+using grpc::ClientWriter;
+using grpc::Status;
+using Core::ProviderMsg;
+using Core::GameCore;
+
+class NetworkInterfaceSubscriber {
+public:
+    virtual void OnEventHappens(const ProviderMsgPtr &) = 0;
+};
+
+class NetworkInterface {
+public:
+    NetworkInterface(const std::shared_ptr<GameCore::StubInterface> &stub) 
+        : mStub(stub)
+    {}
+
+    static std::shared_ptr<GameCore::StubInterface> CreateStub(const std::string &endpoint) {
+        auto channel = grpc::CreateChannel(endpoint, grpc::InsecureChannelCredentials());
+        return GameCore::NewStub(channel);
+    }
+
+    static std::shared_ptr<NetworkInterface> Create(
+        const std::shared_ptr<GameCore::StubInterface> &stub) {
+        return std::make_shared<NetworkInterface>(stub);
+    }
+
+    void Connect() {
+        mStream = mStub->Provider(&mContext);
```

Comment:
could you support client workflow?
client ->  regsiter -> listen.

I think 
1. you can make a pure virtual base class, and offer two implementations. one for provider, another for client.
2. Or, just implement two classes, don't use inheritance.

Reply:
sure, that's under design now